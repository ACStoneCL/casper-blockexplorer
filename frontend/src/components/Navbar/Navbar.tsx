import React, { useState } from 'react';

export interface NavProps {
  readonly label: string | number;
}

const navItems = [
  {
    title: 'Home',
    path: '/',
    id: '0',
  },
  {
    title: 'Deploys',
    path: '/deploys',
    id: '1',
  },
  {
    title: 'Blocks',
    path: '/blocks',
    id: '2',
  },
  {
    title: 'Accounts',
    path: '/accounts',
    id: '3',
  },
];

//accepts arr of item objects
export const Navbar: React.FC = () => {
  const [isOpened, setIsOpened] = useState(false);
  return (
    <nav className="fixed z-10 w-full py-10 pb-10 pr-10 lg:pt-20 lg:pr-56 bg-[#181B38]">
      <div className="flex flex-col lg:flex-row lg:justify-between">
        <div className="flex flex-row justify-end lg:justify-between">
          <div className="lg:hidden" onClick={() => setIsOpened(!isOpened)}>
            {isOpened ? (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                stroke-width="2"
                stroke="white"
                className="w-24 h-24">
                <path
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  d="M6 18L18 6M6 6l12 12"
                />
              </svg>
            ) : (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 24 24"
                fill="white"
                className="w-24 h-24">
                <path
                  fill-rule="evenodd"
                  d="M3 6.75A.75.75 0 013.75 6h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 6.75zM3 12a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 12zm0 5.25a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75a.75.75 0 01-.75-.75z"
                  clip-rule="evenodd"
                />
              </svg>
            )}
          </div>
        </div>
        <div
          onClick={() => setIsOpened(!isOpened)}
          className={`lg:flex lg:space-x-12 lg:flex-row lg:w-auto ${
            isOpened ? 'flex flex-col text-center mt-10 lg:mt-0' : 'hidden'
          }`}>
          {navItems.map((item, index) => {
            return (
              <a
                href={item.path}
                className="text-white py-3 lg:py-0"
                id={item.id}>
                {item.title}
              </a>
            );
          })}
        </div>
      </div>
    </nav>
  );
};
