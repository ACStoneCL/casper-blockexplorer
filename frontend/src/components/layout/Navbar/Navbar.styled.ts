import { Link } from 'react-router-dom';
import styled from '@emotion/styled';

export const Nav = styled.nav`
  @media (min-width: 1024px) {
    width: 25%;
  }
`;

export const NavComponentsContainer = styled.div`
  display: flex;
`;

export const NavButtonContainer = styled.div`
  z-index: 30;
`;

export const NavButton = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: transparent;
  margin: 0 auto;
  padding: 0;
  width: 90%;
  border-style: none;

  @media (min-width: 1024px) {
    display: none;
  }
`;

export const NavItemsContainer = styled.div`
  width: 100%;
`;

export const MobileNav = styled.nav`
  @media (min-width: 1024px) {
    display: none;
  }
`;

export const MobileNavItemsContainer = styled.ul`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100vh;
  width: 100vw;
  align-items: center;
  gap: 1.25rem;
  z-index: 10;
  background-color: #0325d1;
  position: absolute;
  left: 0;
  top: 0;
`;

export const MobileNavItemLink = styled(Link)`
  color: white;
  font-size: 1.375rem;
  padding: 0.3125rem;
  width: 100%;
  font-weight: 500;
  letter-spacing: 0.025em;

  :hover {
    transition: color ease-in 200ms;
    color: #d51e4a;
    transition-property: color;
    text-decoration: none;
  }
`;

export const DesktopNav = styled.nav`
  display: none;

  @media (min-width: 1024px) {
    display: flex;
  }
`;

export const DesktopNavItemsContainer = styled.ul`
  display: flex;
  justify-content: space-between;
  width: 100%;
`;

export const DesktopNavItemLink = styled(Link)`
  color: #0325d1;
  font-size: clamp(0.9rem, 1.2vw, 1.4rem);
  font-weight: 500;
  text-decoration: none;

  :hover,
  :focus {
    background-color: #0325d1;
    transition: ease-in;
    background-image: linear-gradient(
      90deg,
      #1c1e90,
      #693590,
      #d81d54,
      #d81e54,
      #fd6b52
    );
    background-size: 100%;
    -webkit-background-clip: text;
    -moz-background-clip: text;
    -webkit-text-fill-color: transparent;
    -moz-text-fill-color: transparent;
    text-decoration: none;
  }
`;
