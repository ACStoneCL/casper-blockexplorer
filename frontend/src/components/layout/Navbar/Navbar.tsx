import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useAppSelector } from '../../../store';
import { getBounds } from '../../../store';

const navItems = [
  {
    title: 'Home',
    path: '/',
    key: 'home',
  },
  {
    title: 'Peers',
    path: '/peers',
    key: 'peers',
  },
];

export const Navbar: React.FC = () => {
  const [isOpened, setIsOpened] = useState(false);
  const bounds = useAppSelector(getBounds);

  const windowWidth = bounds?.width || 0;

  if (isOpened && windowWidth > 1024) {
    setIsOpened(false);
  }

  useEffect(() => {
    const escKeyHandler = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        event.preventDefault();

        if (isOpened) {
          setIsOpened(false);
        }
      }
    };

    document.addEventListener('keydown', escKeyHandler);

    return () => {
      document.removeEventListener('keydown', escKeyHandler);
    };
  }, [isOpened]);

  return (
    <nav className="z-10 w-full py-10 lg:py-20 bg-[#181B38]">
      <div className="flex flex-col lg:flex-row lg:justify-between">
        <div className="z-30 flex flex-row justify-end lg:justify-between">
          <button
            type="button"
            className="lg:hidden bg-transparent border-none p-0 h-40"
            onClick={() => setIsOpened(!isOpened)}>
            {isOpened ? (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth="2"
                stroke="white"
                className="w-40 h-40">
                <path strokeLinecap="round" d="M6 18L18 6M6 6l12 12" />
              </svg>
            ) : (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 24 24"
                fill="white"
                className="w-40 h-40">
                <path
                  fillRule="evenodd"
                  d="M3 6.75A.75.75 0 013.75 6h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 6.75zM3 12a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 12zm0 5.25a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75a.75.75 0 01-.75-.75z"
                  clipRule="evenodd"
                />
              </svg>
            )}
          </button>
        </div>
        <div className="bg-[#181B38] border-none lg:flex lg:space-x-12 lg:flex-row lg:w-auto">
          {isOpened && (
            <nav className="lg:hidden">
              <ul className="z-10 bg-casper-blue flex flex-col gap-4 absolute w-full h-full items-center justify-center left-0 top-0">
                {navItems.map(({ path, title, key }) => {
                  return (
                    <li key={key}>
                      <Link
                        to={path}
                        className="text-white text-22 py-5 xxs:py-11 w-full font-medium tracking-wide">
                        {title}
                      </Link>
                    </li>
                  );
                })}
              </ul>
            </nav>
          )}
          <nav className="hidden lg:block">
            <ul className="flex gap-x-8">
              {navItems.map(({ path, title, key }) => {
                return (
                  <li key={key}>
                    <Link
                      to={path}
                      className="text-white text-18 py-5 xxs:py-11 lg:py-0 w-full font-medium tracking-wide">
                      {title}
                    </Link>
                  </li>
                );
              })}
            </ul>
          </nav>
        </div>
      </div>
    </nav>
  );
};
