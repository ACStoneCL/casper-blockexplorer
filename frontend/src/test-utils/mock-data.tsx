import { Block } from '../api';

export const mockBlock: Block = {
  height: 1,
  eraID: 1,
  deployCount: 1,
  deployHashes: ['deploy-hash'],
  transferHashes: [],
  timestamp: 123456789,
  readableTimestamp: 'Dec 25th, 1975',
  timeSince: 'Forever Ago..',
  hash: 'block-hash',
  validatorPublicKey: 'validator-public-key',
  parentHash: 'parent-hash',
  stateRootHash: 'state-root-hash',
  rawBlock: '{}',
};

export const createMockBlocks = (numberOfBlocks: number = 10) => {
  const blocks = new Array(numberOfBlocks).fill(mockBlock) as Block[];

  return blocks.map((block, index) => {
    return { ...block, hash: `${block.hash}-${index}` };
  });
};
