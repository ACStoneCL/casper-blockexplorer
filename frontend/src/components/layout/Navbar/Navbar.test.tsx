import React from 'react';
import { render } from '../../../test-utils';
import { Navbar } from './Navbar';

jest.mock('../../../hooks', () => {
  return {
    useAppWidth: () => {
      return { windowWidth: 1022, isMobile: false };
    },
  };
});

describe('Navbar', () => {
  it('should render without error and display nav items', () => {
    const { getByRole, getByText } = render(<Navbar />);

    const nav = getByRole('navigation', { name: 'navigation' });

    expect(nav).toHaveTextContent('Home');

    const navItem = getByText('Home');

    expect(navItem).toBeInTheDocument();
    expect(nav).toBeInTheDocument();
  });

  it('should hide navigation options when screen width is below 1023px', () => {
    const { getByText } = render(<Navbar />);
    const navItem = getByText('Home');

    expect(navItem).not.toBeVisible();
  });
});
