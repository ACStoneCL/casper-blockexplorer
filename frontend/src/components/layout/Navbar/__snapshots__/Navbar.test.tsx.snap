// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should hide navigation options when screen width is below 1023px 1`] = `
<div>
  <nav
    aria-label="navigation"
    class="css-5q49mi"
  >
    <div
      class="css-ov1ktg"
    >
      <div
        class="css-1mbb8hy"
      >
        <div
          class="css-1x9bva"
        >
          <button
            class="css-1r8yndd"
            type="button"
          >
            <svg>
              open-menu-icon.svg
            </svg>
          </button>
        </div>
      </div>
      <div
        class="css-1ago99h"
      >
        <nav
          class="css-1q0v0vc"
        >
          <ul
            class="css-168old1"
          >
            <li>
              <a
                class="css-vz1el3"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                class="css-vz1el3"
                href="/blocks"
              >
                Blocks
              </a>
            </li>
            <li>
              <a
                class="css-vz1el3"
                href="/peers"
              >
                Peers
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </nav>
</div>
`;
